applications/snake-game/snake-game.c

```diff
@@ -0,0 +1,417 @@
+            return;
+        }
+    }
+
+    move_snake(snake_state, next_step);
+
+    if(eatFruit) {
+        place_new_fruit(snake_state);
+    }
+}
+
+void game_logic_thread(void* ctx) {
+    for(;;) {
+        delay(200);
+
+        SnakeState* snake_state = acquire_mutex((ValueMutex*)ctx, 25);
+        if(snake_state == NULL) {
+            continue;
+        }
+        process_game_step(snake_state);
+        release_mutex((ValueMutex*)ctx, snake_state);
+    }
+}
+
+int32_t snake_game_app(void* p) {
+    osMessageQueueId_t event_queue = osMessageQueueNew(8, sizeof(InputEvent), NULL);
+
+    SnakeState snake_state;
+    init_game(&snake_state);
+
+    ValueMutex state_mutex;
+    if(!init_mutex(&state_mutex, &snake_state, sizeof(snake_state))) {
+        printf("cannot create mutex\r\n");
+        return 255;
+    }
+
+    ViewPort* view_port = view_port_alloc();
+    view_port_draw_callback_set(view_port, render_callback, &state_mutex);
+    view_port_input_callback_set(view_port, input_callback, event_queue);
+
+    // Open GUI and register view_port
+    Gui* gui = furi_record_open("gui");
+    gui_add_view_port(gui, view_port, GuiLayerFullscreen);
+
+    // TODO: change to fuirac_start after same in '../music-player/music-player.c'
+    osThreadAttr_t snake_logic_attr = {.name = "snake_logic_thread", .stack_size = 512};
+    osThreadId_t game_logic = osThreadNew(game_logic_thread, &state_mutex, &snake_logic_attr);
+    if(game_logic == NULL) {
+        printf("cannot create player thread\r\n");
+        return 255;
+    }
+
+    InputEvent event;
+    while(1) {
+        osStatus_t event_status = osMessageQueueGet(event_queue, &event, NULL, 100);
+
+        SnakeState* snake_state = (SnakeState*)acquire_mutex_block(&state_mutex);
+
+        if(event_status == osOK) {
+            // press events
+            if(event.type == InputTypeShort) {
+                switch(event.key) {
+                case InputKeyUp:
+                    snake_state->nextMovement = up;
+                    break;
+                case InputKeyDown:
+                    snake_state->nextMovement = down;
+                    break;
+                case InputKeyRight:
+                    snake_state->nextMovement = right;
+                    break;
+                case InputKeyLeft:
+                    snake_state->nextMovement = left;
+                    break;
+                case InputKeyOk:
+                    if(snake_state->state == gameOver) {
+                        init_game(snake_state);
+                    }
+                    break;
+                case InputKeyBack:
+                    release_mutex(&state_mutex, snake_state);
+                    goto exit_loop;
+                }
+            }
+        } else {
+            // event timeout
+        }
+
+        view_port_update(view_port);
+        release_mutex(&state_mutex, snake_state);
+    }
+exit_loop:
+
+    osThreadTerminate(game_logic);
```

Comment:
Terminating threads is forbidden. This process is totally unpredictable and is going to cause mess in value mutex used here.

execution flag + `FuriThread` + `furi_thread_join` is ok, but in this case it is better to switch to timers